File: together-sandbox-typescript/src/Sandboxes.ts
import * as api from "./api-clients/api/index.js";
import {
  createClient as createSandboxClient,
  createConfig as createSandboxConfig,
} from "./api-clients/sandbox/client/index.js";
import { type Client as ApiClient } from "./api-clients/api/client/index.js";
import type { Sandbox as RawSandbox } from "./api-clients/api/types.gen.js";
import { Sandbox } from "./Sandbox.js";
import {
  type SandboxInfo,
  type SandboxStatus,
  type CreateSandboxParams,
  type TerminationSnapshotParams,
  type RetryConfig,
} from "./types.js";
import {
  camelCaseKeys,
  callApi,
  terminationPolicyBody,
  terminationSnapshotBody,
} from "./utils.js";
import { describeLifecycleFailure, isTransientStatus } from "./lifecycle.js";
import { Page } from "./pagination.js";

/**
 * Extract the agent connection details from the Sandbox model.
 */
function resolveConnectionDetails(sandbox: SandboxInfo): {
  url: string;
  token: string;
} {
  if (!sandbox.agent?.url || !sandbox.agent?.token)
    throw new Error("Sandbox has no agent connection details");
  return { url: sandbox.agent.url, token: sandbox.agent.token };
}

/**
 * Wait for a sandbox to reach "running", wire up its client, and return a
 * connected {@link Sandbox}.
 *
 * `sandbox` is the sandbox as last seen (e.g. the createSandbox response). The
 * wait phase is skipped when it has already settled on a non-transient status —
 * waiting would just echo that status back.
 */
async function connectRunningSandbox(
  sandbox: RawSandbox,
  apiClient: ApiClient,
  retryConfig: RetryConfig | undefined,
): Promise<Sandbox> {
  const waitResult = isTransientStatus(sandbox.status)
    ? await callApi(
        "api.waitForSandbox",
        () =>
          api.waitForSandbox({
            client: apiClient,
            path: { id: sandbox.id },
          }),
        retryConfig,
      )
    : sandbox;

  if (waitResult.status !== "running") {
    throw new Error(describeLifecycleFailure(waitResult, "running"));
  }

  const finalData = camelCaseKeys(waitResult);
  const { url, token } = resolveConnectionDetails(finalData);
  const sandboxClient = createSandboxClient(
    createSandboxConfig({
      baseUrl: url,
      headers: { Authorization: `Bearer ${token}` },
    }),
  );

  sandboxClient.interceptors.error.use((error) => error);

  return new Sandbox(finalData, sandboxClient, apiClient, retryConfig);
}

// Default sandbox resource allocation. Match the CLI/snapshot helper.
export const DEFAULT_CPU = 1; // 1 vCPU (cores)
export const DEFAULT_MEMORY_BYTES = 2048 * 1024 * 1024; // 2 GiB

/**
 * Sandbox lifecycle operations, accessed as `sdk.sandboxes.*`.
 */
export class SandboxesNamespace {
  constructor(
    private readonly _apiClient: ApiClient,
    private readonly _retryConfig?: RetryConfig,
  ) {}

  /**
   * Create a sandbox and wait for it to be running, returning a connected {@link Sandbox}.
   *
   * The wait is skipped when the create response already reports a settled
   * status (e.g. the sandbox came back `running`).
   */
  async create(params: CreateSandboxParams = {}): Promise<Sandbox> {
    const data = await callApi(
      "api.createSandbox",
      () =>
        api.createSandbox({
          client: this._apiClient,
          body: {
            snapshot_id: params.snapshotId,
            snapshot_alias: params.snapshotAlias,
            cpu: params.cpu ?? DEFAULT_CPU,
            memory_bytes: params.memoryBytes ?? DEFAULT_MEMORY_BYTES,
            ttl: params.ttl,
            tags: params.tags,
            termination_policy: terminationPolicyBody(params.terminationPolicy) ?? undefined,
          },
        }),
      this._retryConfig,
    );

    return connectRunningSandbox(data, this._apiClient, this._retryConfig);
  }

  /**
   * List sandboxes.
   *
   * Returns a {@link Page} that is async-iterable across all pages — iterate it
   * directly to walk every sandbox, or use `getNextPage()` / `nextCursor` for
   * manual page-by-page control.
   *
   * @param options.limit Max items per page (1–100, default 20).
   * @param options.cursor Resume from a cursor returned by a previous page's
   *   {@link Page.nextCursor} (omit to start from the first page).
   * @param options.statuses Filter by status; matches sandboxes in any of the
   *   given statuses.
   * @param options.snapshotId Filter by snapshot; matches sandboxes created
   *   from the given snapshot.
   * @param options.tags Filter by tags; matches sandboxes whose tags contain
   *   all the given pairs.
   */
  async list(options?: {
    limit?: number;
    cursor?: string;
    statuses?: SandboxStatus[];
    snapshotId?: string;
    tags?: Record<string, string>;
  }): Promise<Page<SandboxInfo>> {
    const fetchPage = async (cursor?: string): Promise<Page<SandboxInfo>> => {
      const result = await callApi(
        "api.listSandboxes",
        () =>
          api.listSandboxes({
            client: this._apiClient,
            query: {
              limit: options?.limit,
              cursor,
              "statuses[]": options?.statuses,
              snapshot_id: options?.snapshotId,
              tags: options?.tags,
            },
          }),
        this._retryConfig,
      );
      return new Page<SandboxInfo>(
        result.data.map((s) => camelCaseKeys(s)),
        result.next_cursor,
        fetchPage,
      );
    };

    return fetchPage(options?.cursor);
  }

  /**
   * Fetch a single sandbox by id. Returns the camelCased {@link SandboxInfo}
   * metadata, consistent with {@link list}.
   */
  async get(sandboxId: string): Promise<SandboxInfo> {
    const data = await callApi(
      "api.getSandbox",
      () =>
        api.getSandbox({
          client: this._apiClient,
          path: { id: sandboxId },
        }),
      this._retryConfig,
      `for sandbox '${sandboxId}'`,
    );

    return camelCaseKeys(data);
  }

  /**
   * Terminate a VM by sandbox ID. After this the sandbox is terminal and cannot
   * be used again.
   *
   * `snapshot` overrides what the sandbox's stored termination policy would
   * snapshot for this teardown (omit to use the stored policy; `null` to make
   * it ephemeral). The produced snapshot is aliased as `sandbox:<sandbox id>`
   * plus any `snapshot.aliases`.
   */
  async terminate(
    sandboxId: string,
    options: {
      snapshot?: TerminationSnapshotParams | null;
    } = {},
  ): Promise<void> {
    await callApi(
      "api.terminateSandbox",
      () =>
        api.terminateSandbox({
          client: this._apiClient,
          path: { id: sandboxId },
          body: {
            snapshot: terminationSnapshotBody(options.snapshot),
          },
        }),
      this._retryConfig,
    );
    const waitResult = await callApi(
      "api.waitForSandbox",
      () =>
        api.waitForSandbox({
          client: this._apiClient,
          path: { id: sandboxId },
        }),
      this._retryConfig,
    );

    if (waitResult.status !== "terminated") {
      throw new Error(describeLifecycleFailure(waitResult, "terminated"));
    }
  }
}

File: together-sandbox-python/together_sandbox/_sandboxes.py
from __future__ import annotations

from .api.client import AuthenticatedClient as ApiClient

from ._sandbox import Sandbox

# ── Management API endpoint functions (detailed variants) ─────────────────────
from .api.api.default.wait_for_sandbox import asyncio_detailed as wait_for_sandbox_api
from .api.api.default.terminate_sandbox import asyncio_detailed as terminate_sandbox_api
from .api.api.default.create_sandbox import asyncio_detailed as create_sandbox_api
from .api.api.default.list_sandboxes import asyncio_detailed as list_sandboxes_api
from .api.api.default.get_sandbox import asyncio_detailed as get_sandbox_api

# ── Management API models ─────────────────────────────────────────────────────
from .api.models.sandbox import Sandbox as SandboxModel
from .api.models.terminate_sandbox_body import TerminateSandboxBody
from .api.models.create_sandbox_body import CreateSandboxBody
from .api.models.tags import Tags
from .api.models.list_sandboxes_statuses_item import ListSandboxesStatusesItem
from .api.types import UNSET, Unset

# ── Helpers ─────────────────────────────────────────────────────
from ._utils import (
    RetryConfig,
    _call_api,
    _resolve_connection,
    build_termination_policy,
    build_termination_snapshot,
    deep_object_tags,
)
from ._pagination import Page
from ._lifecycle import describe_lifecycle_failure, is_transient_status

# ── Sandbox API client ────────────────────────────────────────────────────────
from .sandbox.client import AuthenticatedClient as SandboxClient

# Default sandbox resource allocation. Match the TS SDK / CLI helper.
DEFAULT_CPU = 1.0  # 1 vCPU (cores)
DEFAULT_MEMORY_BYTES = 2048 * 1024 * 1024  # 2 GiB


async def _connect_running_sandbox(
    sandbox: SandboxModel,
    api_client: ApiClient,
    retry: RetryConfig | None,
) -> Sandbox:
    """Wait for a sandbox to reach 'running', wire up its client, and return it.

    ``sandbox`` is the sandbox as last seen (e.g. the create_sandbox response).
    The wait phase is skipped when it has already settled on a non-transient
    status — waiting would just echo that status back.

    Used by :meth:`SandboxesNamespace.create`.
    """
    sandbox_id = sandbox.id
    vm_info: SandboxModel = sandbox

    if is_transient_status(sandbox.status):
        vm_info = await _call_api(
            "api.wait_for_sandbox",
            lambda: wait_for_sandbox_api(sandbox_id, client=api_client),
            retry,
            context=f"for sandbox {sandbox_id!r}",
        )

    if vm_info.status != "running":
        raise RuntimeError(describe_lifecycle_failure(vm_info, "running"))

    url, token = _resolve_connection(vm_info)

    sandbox_client = SandboxClient(
        base_url=url,
        token=token,
        prefix="Bearer",
    )

    return Sandbox(vm_info, sandbox_client, api_client, retry=retry)


class SandboxesNamespace:
    """Sandbox lifecycle operations accessed as ``sdk.sandboxes.*``."""

    def __init__(
        self,
        api_client: ApiClient,
        *,
        retry: RetryConfig | None = None,
    ) -> None:
        self._api_client = api_client
        self._retry = retry

    async def create(
        self,
        *,
        cpu: float = DEFAULT_CPU,
        memory_bytes: int = DEFAULT_MEMORY_BYTES,
        snapshot_id: str | None = None,
        snapshot_alias: str | None = None,
        ttl: int | None = None,
        tags: dict[str, str] | None = None,
        termination_policy: dict | None = None,
    ) -> Sandbox:
        """Create a sandbox and wait for it to be running.

        The wait is skipped when the create response already reports a settled
        status (e.g. the sandbox came back ``running``).

        Args:
            cpu: CPU allocation in cores (e.g. 1 = 1 vCPU). Must be between 0.1 and 16.
            memory_bytes: Memory allocation in bytes. Must be between 1 GB and
                8 GB per requested CPU.
            snapshot_id: Optional snapshot ID to create the sandbox from.
            snapshot_alias: Optional snapshot alias to create the sandbox from.
            ttl: Optional seconds after creation before the sandbox is
                automatically terminated.
            tags: Optional key/value labels to attach to the sandbox.
            termination_policy: The termination snapshot policy, e.g.
                ``{"snapshot": {"aliases": ["prod"]}}``.
                Omit for an ephemeral sandbox (no snapshot, deleted on termination).

        """
        body = CreateSandboxBody(
            snapshot_id=snapshot_id if snapshot_id is not None else UNSET,
            snapshot_alias=snapshot_alias if snapshot_alias is not None else UNSET,
            cpu=cpu,
            memory_bytes=memory_bytes,
            ttl=ttl if ttl is not None else UNSET,
            tags=Tags.from_dict(tags) if tags is not None else UNSET,
            termination_policy=build_termination_policy(termination_policy),
        )
        sandbox_model: SandboxModel = await _call_api(
            "api.create_sandbox",
            lambda: create_sandbox_api(client=self._api_client, body=body),
            self._retry,
        )

        return await _connect_running_sandbox(sandbox_model, self._api_client, self._retry)

    async def list(
        self,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        statuses: list[str] | None = None,
        tags: dict[str, str] | None = None,
    ) -> Page[SandboxModel]:
        """List sandboxes.

        Returns a :class:`Page` that is async-iterable across all pages —
        iterate it directly to walk every sandbox, or use ``get_next_page()``
        / ``next_cursor`` for manual page-by-page control.

        Args:
            limit: Max items per page (1–100, default 20).
            cursor: A ``next_cursor`` value returned by a previous page (omit
                to start from the first page).
            statuses: Filter by status; matches sandboxes in any of the given
                statuses.
            tags: Filter by tags; matches sandboxes whose tags contain all the
                given pairs.

        Returns:
            Page[Sandbox]: First page of sandboxes.

        Example:
            >>> async for sandbox in await sdk.sandboxes.list(statuses=["running"]):
            ...     print(sandbox.id)
        """

        async def fetch_page(cursor: str | None = None) -> Page[SandboxModel]:
            result = await _call_api(
                "api.list_sandboxes",
                lambda: list_sandboxes_api(
                    client=self._api_client,
                    limit=limit if limit is not None else UNSET,
                    cursor=cursor if cursor is not None else UNSET,
                    statuses=(
                        [ListSandboxesStatusesItem(s) for s in statuses]
                        if statuses is not None
                        else UNSET
                    ),
                    tags=deep_object_tags(tags),
                ),
                self._retry,
            )
            return Page(result.data, result.next_cursor, fetch_page)

        return await fetch_page(cursor)

    async def get(self, sandbox_id: str) -> SandboxModel:
        """Fetch a single sandbox by id.

        Returns the raw sandbox metadata, consistent with :meth:`list` — not
        a wired :class:`Sandbox` client.

        Args:
            sandbox_id: The sandbox to fetch.

        Returns:
            Sandbox: The sandbox metadata record.

        Example:
            >>> sandbox = await sdk.sandboxes.get("sb_1")
            >>> print(sandbox.status)
        """
        return await _call_api(
            "api.get_sandbox",
            lambda: get_sandbox_api(sandbox_id, client=self._api_client),
            self._retry,
            context=f"for sandbox {sandbox_id!r}",
        )

    async def terminate(
        self,
        sandbox_id: str,
        *,
        snapshot: dict | None | Unset = UNSET,
    ) -> None:
        """Terminate a VM by sandbox ID.

        After this the sandbox is terminal and cannot be used again.

        Args:
            sandbox_id: The sandbox to terminate.
            snapshot: What this teardown snapshots, overriding the snapshot the
                sandbox's stored termination policy would take, e.g.
                ``{"aliases": ["prod"]}``. Omit (the default) to keep the stored
                policy; pass ``None`` to make the teardown ephemeral (no
                snapshot). The produced snapshot is aliased as
                ``sandbox:<sandbox id>`` plus any ``aliases``.
        """
        await _call_api(
            "api.terminate_sandbox",
            lambda: terminate_sandbox_api(
                sandbox_id,
                client=self._api_client,
                body=TerminateSandboxBody(
                    snapshot=build_termination_snapshot(snapshot),
                ),
            ),
            self._retry,
            context=f"for sandbox {sandbox_id!r}",
        )

        vm_info: SandboxModel = await _call_api(
            "api.wait_for_sandbox",
            lambda: wait_for_sandbox_api(sandbox_id, client=self._api_client),
            self._retry,
            context=f"for sandbox {sandbox_id!r}",
        )

        if vm_info.status != "terminated":
            raise RuntimeError(describe_lifecycle_failure(vm_info, "terminated"))

